src/nuklear_knob.c

```diff
@@ -0,0 +1,250 @@
+#include "nuklear.h"
+#include "nuklear_internal.h"
+
+/* ===============================================================
+ *
+ *                               KNOB
+ *
+ * ===============================================================*/
+
+NK_LIB float
+nk_knob_behavior(nk_flags *state, struct nk_input *in,
+    struct nk_rect bounds, float knob_min, float knob_max, float knob_value,
+    float knob_step, float knob_steps,
+    enum nk_heading zero_direction, float dead_zone_percent)
+{
+    struct nk_vec2 origin;
+    float angle = 0.0f;
+    origin.x = bounds.x + (bounds.w / 2);
+    origin.y = bounds.y + (bounds.h / 2);
+
+    nk_widget_state_reset(state);
+
+    /* handle click and drag input */
+    if(in &&
+       in->mouse.buttons[NK_BUTTON_LEFT].down &&
+       nk_input_has_mouse_click_down_in_rect(in, NK_BUTTON_LEFT, bounds, nk_true)){
+        /* calculate angle from origin and rotate */
+        const float direction_rads[4] = {
+            NK_PI * 2.5f, /* 90  NK_UP */
+            NK_PI * 2.0f, /* 0   NK_RIGHT */
+            NK_PI * 1.5f, /* 270 NK_DOWN */
+            NK_PI,        /* 180 NK_LEFT */
+        };
+        *state = NK_WIDGET_STATE_ACTIVE;
+
+        angle = NK_ATAN2(in->mouse.pos.y - origin.y, in->mouse.pos.x - origin.x) + direction_rads[zero_direction];
+        angle -= (angle > NK_PI * 2) ? NK_PI * 3 : NK_PI;
+
+        /* account for dead space applied when drawing */
+        angle *= 1.0f / (1.0f - dead_zone_percent);
+        angle = NK_CLAMP(-NK_PI, angle, NK_PI);
+
+        /* convert -pi -> pi range to 0.0 -> 1.0 */
+        angle = (angle + NK_PI) / (NK_PI * 2);
+
+        /* click to closest step */
+        knob_value = knob_min + ( (int)(angle * knob_steps + (knob_step / 2)) ) * knob_step;
+        knob_value = NK_CLAMP(knob_min, knob_value, knob_max);
+    }
+
+    /* knob widget state */
+    if (nk_input_is_mouse_hovering_rect(in, bounds)){
+        *state = NK_WIDGET_STATE_HOVERED;
+        if (in) {
+            /* handle scroll and arrow inputs */
+            if (in->mouse.scroll_delta.y > 0 ||
+               (in->keyboard.keys[NK_KEY_UP].down && in->keyboard.keys[NK_KEY_UP].clicked))
```

Comment:
Mouse hovering with key state changes isn't a user interaction behavior across any of the other widgets. I'm conflicted on this one.